integration/container/stop_linux_test.go

```diff
@@ -64,3 +67,66 @@ func TestStopContainerWithTimeout(t *testing.T) {
 		})
 	}
 }
+
+// TestStopContainerWithTimeoutCancel checks that ContainerStop is not cancelled
+// if the request is cancelled or the connection is closed.
+// See issue https://github.com/moby/moby/issues/45731
+func TestStopContainerWithTimeoutCancel(t *testing.T) {
+	t.Parallel()
+	defer setupTest(t)()
+
+	const (
+		cancelContext   = "cancel context"
+		closeConnection = "close connection"
+	)
+
+	for _, tc := range []string{cancelContext, closeConnection} {
+		tc := tc
+		t.Run(tc, func(t *testing.T) {
+			t.Parallel()
+			apiClient := testEnv.APIClient()
+			defer apiClient.Close()
+			ctx := context.Background()
+
+			testCmd := container.WithCmd("sh", "-c", "sleep 30")
+			id := container.Run(ctx, t, apiClient, testCmd)
+
+			ctxCancel, cancel := context.WithCancel(context.Background())
+			stopTimeout := 3
+			defer cancel()
+			go func() {
+				err := apiClient.ContainerStop(ctxCancel, id, containertypes.StopOptions{Timeout: &stopTimeout})
+				switch tc {
+				case cancelContext:
+					assert.Check(t, is.ErrorType(err, errdefs.IsCancelled))
+				case closeConnection:
+					assert.Check(t, err)
+				}
+			}()
+
+			// Give the ContainerStop some time to make sure it's being handled,
+			// then cancel the context or close the client to cancel the stop.
+			time.Sleep(100 * time.Millisecond)
```

Comment:
Not very happy with the test (and this sleep); better suggestions welcome 😅

Reply:
Collude with the container: trap the stop signal and handle it by logging to stdout then sleeping some more, similar to the tests in wait_test.go. The test can then wait for the log message from the container to know when it can proceed.